Create nft.transfers: a table combining erc721, erc1155single & erc1155batch transfers

hildobby opened the pull request:
Brief comments on the purpose of your changes:

Combining all nft transfer tables to more easily access all nft transfers with 1 join instead of 3 different joins

*For Dune Engine V2*
I've checked that:
General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)

Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

hildobby commented:
Seeing this error:
```
Model 'model.spellbook.nft_avalanche_c_transfers' (models/nft/avalanche_c/nft_avalanche_c_transfers.sql) depends on a source named 'erc721_avalanche_c.evt_Transfer' which was not found
```

But the table erc721_avalanche_c.evt_Transfer does exist so I don't get it 🤷 
Any idea what the error is here?

Hosuke commented:
> Seeing this error:
> 
> ```
> Model 'model.spellbook.nft_avalanche_c_transfers' (models/nft/avalanche_c/nft_avalanche_c_transfers.sql) depends on a source named 'erc721_avalanche_c.evt_Transfer' which was not found
> ```
> 
> But the table `erc721_avalanche_c.evt_Transfer` does exist so I don't get it 🤷 Any idea what the error is here?

Hi @hildobby , I think the problem about 'erc721_avalanche_c.evt_Transfer' was not found may be caused by this table is not defined in any sources.yml file within the project. We may need to add this table into [base sources](https://github.com/duneanalytics/spellbook/blob/main/models/base_sources/avalanche_c_base_sources.yml) or the model source, probably naming `models/nft/avalanche_c/nft_avalanche_c_sources.yml`

hildobby commented:
> Hi @hildobby , I think the problem about `erc721_avalanche_c.evt_Transfer` not found may be caused by this table is not defined in any sources.yml file within the project. We may need to add this table into [base sources](https://github.com/duneanalytics/spellbook/blob/main/models/base_sources/avalanche_c_base_sources.yml) or the model source, probably naming `models/nft/avalanche_c/nft_avalanche_c_sources.yml`

thx for the feedback! I added all erc20, erc721 & erc1155(single&batch) contracts for each chain into base sources to match how it is for ethereum tables. I also rectified a few bugs I had.

I encountered another error however and I'm unsure what the root cause is 🤔

Hosuke commented:
Hi @hildobby , so what's the current error at the moment? Let me see if I can help with it.
For [dbt slim cli test failure](https://github.com/duneanalytics/spellbook/actions/runs/3388193587/jobs/5629878615), I think it links to the config block (usually at beginning of the sql file, and we define unique keys, partition method, incremental strategy and post hooks there) in the model sql. You can try to add something like [this block](https://github.com/duneanalytics/spellbook/blob/main/models/sushiswap/ethereum/sushiswap_ethereum_trades.sql#L1#L14) in your .sql files.

hildobby commented:
> Hi @hildobby , so what's the current error at the moment? Let me see if I can help with it. For [dbt slim cli test failure](https://github.com/duneanalytics/spellbook/actions/runs/3388193587/jobs/5629878615), I think it links to the config block (usually at beginning of the sql file, and we define unique keys, partition method, incremental strategy and post hooks there) in the model sql. You can try to add something like [this block](https://github.com/duneanalytics/spellbook/blob/main/models/sushiswap/ethereum/sushiswap_ethereum_trades.sql#L1#L14) in your .sql files.

Thanks for the help I fixed the issue, the tests are now passed except for the incremental run that times out due to the 1h timeout limit. So it should be good to go right?

Hosuke commented on models/nft/avalanche_c/nft_avalanche_c_schema.yml:
```diff
@@ -15,3 +15,45 @@ models:
+      contributors: hildobby
+    config:
+      tags: ['nft', 'avalanche_c', 'transfers']
+    description: >
+        NFT transfers
+    columns:
+      - name: block_time
+        description: "UTC event block time"
+      - name: block_number
```
You can add a uniqueness test here if you wish, currently the unique key only consists of one column `unique_transfer_id` but there may be a list of columns that form a unique key in the future.
```suggestion
    description: >
        NFT transfers
    tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - unique_transfer_id
    columns:
      - name: block_time
        description: "UTC event block time"
      - name: block_number
```

Hosuke commented on models/nft/avalanche_c/nft_avalanche_c_transfers.sql:
```diff
@@ -0,0 +1,62 @@
+ {{ config( alias='transfers') }}
+
```
SInce nft_transfers views are updated incrementally by chain(as shown in dbt_project.yml),
we can fill more information in the config blocks for models under different chains.
```suggestion
{{ config(
        alias ='transfers',
        partition_by='block_date',
        materialized='incremental',
        file_format = 'delta',
        post_hook='{{ expose_spells(\'["avalanche_c"]\',
                                    "sector",
                                    "nft",
                                    \'["hildobby"]\') }}',
        unique_key = ['unique_transfer_id']
)
}}
```

Hosuke commented on models/nft/bnb/nft_bnb_transfers.sql:
```diff
@@ -0,0 +1,62 @@
+ {{ config( alias='transfers') }}
```
Same as above.
```suggestion
{{ config(
        alias ='transfers',
        partition_by='block_date',
        materialized='incremental',
        file_format = 'delta',
        post_hook='{{ expose_spells(\'["bnb"]\',
                                    "sector",
                                    "nft",
                                    \'["hildobby"]\') }}',
        unique_key = ['unique_transfer_id']
)
}}
```

Hosuke commented on models/nft/gnosis/nft_gnosis_transfers.sql:
```diff
@@ -0,0 +1,62 @@
+ {{ config( alias='transfers') }}
```
Same as above.
```suggestion
{{ config(
        alias ='transfers',
        partition_by='block_date',
        materialized='incremental',
        file_format = 'delta',
        post_hook='{{ expose_spells(\'["gnosis"]\',
                                    "sector",
                                    "nft",
                                    \'["hildobby"]\') }}',
        unique_key = ['unique_transfer_id']
)
}}
```

Hosuke commented on models/nft/nft_transfers.sql:
```diff
@@ -0,0 +1,124 @@
+, unique_transfer_id
+FROM {{ ref('nft_bnb_transfers') }}
+UNION
+SELECT 'avalanche' AS blockchain
```
I am not sure if we need to use 'avalanche_c' for the blockchain(as we did in the prices.usd table).
```suggestion
SELECT 'avalanche_c' AS blockchain
```

Hosuke commented on models/nft/nft_transfers.sql:
```diff
@@ -0,0 +1,124 @@
+        partition_by='block_date',
+        materialized='incremental',
+        file_format = 'delta',
+        post_hook='{{ expose_spells(\'["ethereum"]\',
```
And here becomes multi-chains.
```suggestion
        post_hook='{{ expose_spells(\'["ethereum", "bnb", "avalanche_c", "gnosis", "optimism", "arbitrum", "polygon"]\',
```

Hosuke commented on models/nft/ethereum/nft_ethereum_transfers.sql:
```diff
@@ -0,0 +1,62 @@
+ {{ config( alias='transfers') }}
```
Same as above.
```suggestion
{{ config(
        alias ='transfers',
        partition_by='block_date',
        materialized='incremental',
        file_format = 'delta',
        post_hook='{{ expose_spells(\'["ethereum"]\',
                                    "sector",
                                    "nft",
                                    \'["hildobby"]\') }}',
        unique_key = ['unique_transfer_id']
)
}}
```

Hosuke commented on models/nft/polygon/nft_polygon_transfers.sql:
```diff
@@ -0,0 +1,62 @@
+ {{ config( alias='transfers') }}
```
Same as above.
```suggestion
{{ config(
        alias ='transfers',
        partition_by='block_date',
        materialized='incremental',
        file_format = 'delta',
        post_hook='{{ expose_spells(\'["polygon"]\',
                                    "sector",
                                    "nft",
                                    \'["hildobby"]\') }}',
        unique_key = ['unique_transfer_id']
)
}}
```

Hosuke commented on models/nft/optimism/nft_optimism_transfers.sql:
```diff
@@ -0,0 +1,62 @@
+ {{ config( alias='transfers') }}
```
Same as above.
```suggestion
{{ config(
        alias ='transfers',
        partition_by='block_date',
        materialized='incremental',
        file_format = 'delta',
        post_hook='{{ expose_spells(\'["optimism"]\',
                                    "sector",
                                    "nft",
                                    \'["hildobby"]\') }}',
        unique_key = ['unique_transfer_id']
)
}}
```

Hosuke reviewed (changes requested):
Since the nft_transfers materialized view is updated incrementally by chain, I think we need to add information in the config blocks for each model.

Moreover it will be great if we add some seeds to these models to ensure the correctness of the data.

hildobby commented:
ty again for the help @Hosuke! I agree with all your suggestions so I committed those. I also added anti joins on tx_hash  against itself so that incremental runs don't lead to duplicates. I also switched UNION joins to UNION ALLs which should lead to faster model runs

Hosuke commented on models/nft/arbitrum/nft_arbitrum_transfers.sql:
```diff
@@ -0,0 +1,74 @@
+ {{ config( alias='transfers') }}
```
This would be the last config block that need to add incremental strategy in this PR.

hildobby replied:
ah well spotted I had missed that one, just fixed it with a new commit :)

Hosuke commented:
Since I am relatively less experienced in nft.trades, I think I would pass this PR to @soispoke to take a look.
Here is the summary of the parts I have reviewed:
- Metadata & naming conventions look good ✅
- Incremental filters added to every CTE when querying large tables ✅
- Contract addresses are lowercased ✅
- Config blocks added for incremental views ✅

Hosuke commented on models/nft/arbitrum/nft_arbitrum_transfers.sql:
```diff
@@ -0,0 +1,85 @@
+, 'arbitrum' || t.evt_tx_hash || '-erc721-' || t.contract_address || '-' || t.tokenId || '-' || t.from || '-' || t.to || '-' || '1' || '-' || t.evt_index AS unique_transfer_id
+FROM {{ source('erc721_arbitrum','evt_transfer') }} t
+    {% if is_incremental() %}
+        LEFT ANTI JOIN {{this}} anti
```
For [test failure](https://github.com/duneanalytics/spellbook/actions/runs/3410941176/jobs/5674523685), you can retry with [ANTI JOIN](https://docs.databricks.com/sql/language-manual/sql-ref-syntax-qry-select-join.html).
```suggestion
        ANTI JOIN {{this}} anti
```

hildobby replied:
Sure just changed it in a new commit, but why do you think that would work better? Here it worked fine with a `LEFT ANTI JOIN` but maybe I'm missing something: https://github.com/duneanalytics/spellbook/blob/main/models/addresses_events/ethereum/addresses_events_ethereum_first_funded_by.sql

hildobby replied:
It seems like that didn't fix the issue 😔 Not sure what the problem is since I had this exact method worked on file linked above 🤷  Lmk if you have any ideas!

Hosuke replied:
It seems like the one you wrote is the first `ANTI JOIN` in the spellbook, and there is no other precedent to refer to at the moment. Maybe try this:
```
LEFT JOIN {{this}} anti
ON t.evt_tx_hash = anti.tx_hash
AND anti.tx_hash is NULL
```

Hosuke replied:
So I replace `anti` to `anti_table` and solve the syntax error. I think the word `anti` is a keyword in the databricks sql.

Hosuke commented:
The code takes more than 60 minutes to run dbt-test, so currently it may need to be tested/checked manually

soispoke commented:
@Hosuke @hildobby Looking into this, test is failing but on an unrelated model (@chuxinh I think that's yours :) )

soispoke commented:
Local tests look good, waiting for dbt slim ci to restart but will go ahead and merge after

soispoke commented:
wait I almost missed this but @hildobby you should add incremental filters on the final nft_transfers model 🙏

augustog commented:
Hey @hildobby, this is an incredible addition, thanks for this ser <3

I'm looking into how we can update our nft models with reference to this table (Probably not a simple task). What's the intended joining key to replace the current three joins?

hildobby commented:
Sorry just saw your comment @augustog. imo `tx_hash`, `contract_address` & `token_id` (& optional `block_time` for optimisation) should suffice for most use cases. I did also purposely leave `evt_index` that could prove useful for some joins :)
